File: src/main/scala/DPhyp/PowerSet.java
package DPhyp;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.Iterator;

public class PowerSet implements Iterator<BitSet>,Iterable<BitSet>{
    private ArrayList<Integer> arr = null;
    private BitSet bset = null;

    @SuppressWarnings("unchecked")
    public PowerSet(BitSet set)
    {
    	arr = new ArrayList<Integer>();
        for (int i = 0; i <= set.size(); i++) {
        	if(set.get(i))
        		arr.add(i);
		}
        
        bset = new BitSet(arr.size() + 1);
    }

    @Override
    public boolean hasNext() {
        return !bset.get(arr.size());
    }

    @Override
    public BitSet next() {
    	BitSet returnSet = new BitSet();
        for(int i = 0; i < arr.size(); i++)
        {
            if(bset.get(i))
                returnSet.set(arr.get(i));
        }
        //increment bset
        for(int i = 0; i < bset.size(); i++) 
        {
            if(!bset.get(i))
            {
                bset.set(i);
                break;
            }else
                bset.clear(i);
        }

        return returnSet;
    }

    @Override
    public void remove() {
        throw new UnsupportedOperationException("Not Supported!");
    }

    @Override
    public Iterator<BitSet> iterator() {
        return this;
    }
    
    public static void main(String[] args) {
    	long time =System.currentTimeMillis();
    	int n =3;
		BitSet b = new BitSet(n);
		for (int i = 1; i <= n; i++) {
			b.set(i);
		}
		int count =0;
		PowerSet p = new PowerSet(b);
		for(BitSet b1 : p){
			System.out.println(b1);
			count++;
		}
		System.out.println("Count: "+count);
    	long stoptime =System.currentTimeMillis();
    	System.out.println("Time ms: "+(stoptime-time));
	}
}

File: src/main/scala/gr/cslab/ece/ntua/musqle/plan/hypergraph/PowerSet.scala
package gr.cslab.ece.ntua.musqle.plan.hypergraph

import java.util

import scala.util.control.Breaks._

class PowerSet(set: util.BitSet) {
  private final val arr: util.ArrayList[Integer] = new util.ArrayList[Integer]()

  for (i <- Range(0, set.size)){
    if (set.get(i)){arr.add(i)}
  }
  private final val bitSet: util.BitSet = new util.BitSet(arr.size + 1)

  @Override def hasNext: Boolean = !bitSet.get(arr.size)

  @Override def next: util.BitSet = {
    val returnSet = new util.BitSet()
    for(i <- Range(0, arr.size())){
      if(bitSet.get(i))
        returnSet.set(arr.get(i));
    }
    //increment bitSet
    breakable {
      for(i <- Range(0, bitSet.size())) {
        if (!bitSet.get(i)) {
          bitSet.set(i)
          break
        }
        else {
          bitSet.clear(i)
        }
      }
    }

    returnSet
  }

  @Override def remove() {
    throw new Exception("Not Supported!")
  }

  @Override def iterator: PowerSet = this
}
